awx_collection/plugins/modules/tower_credential.py

```diff
@@ -257,162 +274,117 @@
-                params['description'] = module.params.get('description')
-
-            if module.params.get('user'):
-                user_res = tower_cli.get_resource('user')
-                user = user_res.get(username=module.params.get('user'))
-                params['user'] = user['id']
-
-            if module.params.get('team'):
-                team_res = tower_cli.get_resource('team')
-                team = team_res.get(name=module.params.get('team'))
-                params['team'] = team['id']
-
-            if module.params.get('ssh_key_data'):
-                data = module.params.get('ssh_key_data')
-                if os.path.exists(data):
-                    module.deprecate(
-                        msg='ssh_key_data should be a string, not a path to a file.',
-                        version="2.12"
-                    )
-                    if os.path.isdir(data):
-                        module.fail_json(msg='attempted to read contents of directory: %s' % data)
-                    with open(data, 'rb') as f:
-                        module.params['ssh_key_data'] = to_text(f.read())
-                else:
-                    module.params['ssh_key_data'] = data
-
-            if module.params.get('vault_id', None) and module.params.get('kind') != 'vault':
-                module.fail_json(msg="Parameter 'vault_id' is only valid if parameter 'kind' is specified as 'vault'")
-
-            for key in OLD_INPUT_NAMES:
-                if 'kind' in params:
-                    params[key] = module.params.get(key)
-                elif module.params.get(key):
-                    params.setdefault('inputs', {})[key] = module.params.get(key)
-
-            if state == 'present':
-                result = credential.modify(**params)
-                json_output['id'] = result['id']
-            elif state == 'absent':
-                result = credential.delete(**params)
-        except (exc.NotFound) as excinfo:
-            module.fail_json(msg='Failed to update credential, organization not found: {0}'.format(excinfo), changed=False)
-        except (exc.ConnectionError, exc.BadRequest, exc.AuthError) as excinfo:
-            module.fail_json(msg='Failed to update credential: {0}'.format(excinfo), changed=False)
-
-    json_output['changed'] = result['changed']
-    module.exit_json(**json_output)
+    # Attempt to look up the related items the user specified (these will fail the module if not found)
+    if organization:
+        org_id = module.resolve_name_to_id('organizations', organization)
+    if user:
+        user_id = module.resolve_name_to_id('users', user)
+    if team:
+        team_id = module.resolve_name_to_id('teams', team)
+
+    if kind:
+        module.deprecate(msg='The kind parameter has been depricated, please use credential_type instead', version="3.6")
+
+    cred_type_id = module.resolve_name_to_id('credential_types', credential_type if credential_type else KIND_CHOICES[kind])
```

Comment:
Ugly nuance - Credential Types are not unique on name.

https://github.com/ansible/awx/blob/0362c88e48f9862fb5d33da594faeec2b1327d59/awx/main/models/credential/__init__.py#L311-L322

By itself, this uniqueness rule was a little confusing, but integrating it correctly into any client library like this completely blows up any code structure someone was trying to use. It's made worse by the fact that it is truly an exception. There is literally no other model that defines a uniqueness rule on 2 character fields (common is 1 character field and 1 foreign key, which then recurses in named URL).

Had to get that out of the way. So pragmatically, do we solve it here? Do we punt it as something to solve later?

Reply:
current plan:

as we merge this I will file an issue. I might frame it more-so as an API issue as opposed to with the collection.